### PR TITLE
bpo-31429: Define TLS cipher suite on build time 

### PR DESCRIPTION
**Pull request by tiran**
Until now Python used a hard coded white list of default TLS cipher
suites. The old approach has multiple downsides. OpenSSL's default
selection was completely overruled. Python did neither benefit from new
cipher suites (ChaCha20, TLS 1.3 suites) nor blacklisted cipher suites.
For example we used to re-enable 3DES.

Python now defaults to OpenSSL DEFAULT cipher suite selection and black
lists all unwanted ciphers. Downstream vendors can override the default
cipher list with --with-ssl-default-suites.

Signed-off-by: Christian Heimes <christian@python.org>

https://bugs.python.org/issue31429

### CODE REVIEW

**Comment by encukou** [comment]
Is there any way I can help move this forward?

**Comment by tiran** [comment]
I'm not yet sure this patch can land in 3.7. LibreSSL is messing up my plans.

**Comment by tiran** [comment]
This PR can be merged after #3462 has landed or we forcefully unsupport OpenSSL < 1.0.2.

File: Lib/ssl.py
Comment (alex) [inline Lib/ssl.py]: You can just remove this `elif` block entirely :-)

File: Misc/NEWS.d/next/Build/2017-09-13-08-52-58.bpo-31429._vO3LJ.rst
Comment (alex) [inline Misc/NEWS.d/next/Build/2017-09-13-08-52-58.bpo-31429._vO3LJ.rst]: Probably needs the WIP removed :-)

**Comment by tiran** [comment]
@ned-deily Fedora upstream would love have this PR in 3.7. It's also going to make your life as 3.7 security maintainer easier.

@alex @njsmith @pitrou Please review

File: Modules/_ssl.c
Comment (alex) [inline Modules/_ssl.c]: Is this still needed after the other change to require 1.0.2?

**Review by alex — COMMENTED** [review]
Looks ok minus these comments, but I'm not really qualified to review configure changes.

File: Lib/ssl.py
Comment (tiran) [reply Lib/ssl.py]: Done

File: Misc/NEWS.d/next/Build/2017-09-13-08-52-58.bpo-31429._vO3LJ.rst
Comment (tiran) [reply Misc/NEWS.d/next/Build/2017-09-13-08-52-58.bpo-31429._vO3LJ.rst]: Oh, I forgot to update the old blurb. I have created a new one and also added 3.7 whatsnew.

File: Modules/_ssl.c
Comment (tiran) [reply Modules/_ssl.c]: Not really, I added it for doubleplusgood paranoia reasons and just removed it.

**Review by pitrou — APPROVED** [review]
This looks both fine and useful to me.

**Pull request merged**
